spec/rspec/core/reentrant_mutex_spec.rb

```diff
@@ -0,0 +1,37 @@
+require 'rspec/core/reentrant_mutex'
+
+# There are no assertions specifically
+# They are pass if they don't deadlock
+RSpec.describe RSpec::Core::ReentrantMutex do
+  let!(:mutex) { described_class.new }
+  let!(:order) { ThreadOrderSupport.new }
+  after { order.apocalypse! }
+
+  it 'can repeatedly synchronize within the same thread' do
+    mutex.synchronize { mutex.synchronize { } }
+  end
+
+  it 'locks other threads out while in the synchronize block' do
+    order.declare(:before) { mutex.synchronize { } }
+    order.declare(:within) { mutex.synchronize { } }
+    order.declare(:after)  { mutex.synchronize { } }
+
+    order.pass_to :before, :resume_on => :exit
+    mutex.synchronize { order.pass_to :before, :resume_on => :sleep }
+    order.pass_to :before, :resume_on => :exit
+  end
+
+  it 'resumes the next thread once all its synchronize blocks have completed' do
+    order.declare(:thread) { mutex.synchronize { } }
+    mutex.synchronize { order.pass_to :thread, :resume_on => :sleep }
+    order.apocalypse! :join
+  end
+
+  it 'is implemented without depending on the stdlib' do
+    loaded_filenames = $LOADED_FEATURES.map { |filepath| File.basename filepath }
+    pending 'thread seems to be required from core, and something is still requiring monitor'
+    expect(loaded_filenames).to_not include 'monitor.rb'
+    expect(loaded_filenames).to_not include 'thread.rb'
+    expect(loaded_filenames).to_not include 'thread.bundle'
+  end
```

Comment:
> I see. The downside of this, though, is that RSpec could could use those libs unintentionally and pass b/c they're available in its test env, but not have them available in the user's test env. Unless I'm just still confused about how it works.

I think our cukes guard against this -- each scenario writes out a file, runs it via the `rspec` command and has expectations on the produced output.  When the cuke does that, it's shelling out and running a new ruby process and booting rspec from scratch to run the one file -- so it has no pollution from any of our test environment stuff.  We don't exhaustively test every code path via cukes obviously (that would be horrible) but they cover a large percentage of the features, enough that I'm not concerned about this in practice.